docs/jep/11759-python-package-structure.md

```diff
@@ -0,0 +1,142 @@
+    API implementation. This wheel would be built once for each OS (e.g.,
+    `jaxlib-cpu-cp3-abi3-manylinux2014_x86_64.whl`).
+
+    We split `jaxlib-cpu` from `jaxlib` to reduce the size of the binary
+    distribution and build time: with this split, we need only build XLA/LLVM
+    for CPU once, rather than once per Python version.
+
+    We require that the versions of `jaxlib` and `jaxlib-cpu` match exactly.
+
+*   `jaxlib-cuda-11-1`, `jaxlib-cuda-11-4`, `jaxlib-rocm-...` etc.: wheels
+    containing GPU support as optional plugins for the `jaxlib` wheel. These
+    wheels would be built for the product of {CUDA/ROCm versions} x {OSes}
+    (e.g., `jaxlib-cuda-11-1-0.3.14-cp3-abi3-manylinux2014_x86_64.whl`).
+
+    We require that the versions of `jaxlib` and `jaxlib-cuda-...` match
+    exactly.
+
+    These wheels would also be structured to avoid a minor Python version
+    dependency, i.e., the `cuda` should work for all Python 3 versions we
+    support, rather than needing to be, say, Python 3.7 specific. This means
+    that we can only target the Python stable C API and must avoid the use of
+    `pybind11`.
+
+    It is to be determined whether we need only a single `jaxlib-cuda` plugin or
+    if we need multiple CUDA releases, however that is largely orthogonal to
+    this proposal.
+
+*   `jaxlib-tpu`: wheels containing TPU support as an optional plugin for the
+    `jaxlib` wheel. This wheel would be built only for Linux Cloud TPU VMs
+    (e.g., `jaxlib-tpu-0.3.14-cp3-abi3-linux_x86_64.whl`).
+
+
+## Why do this?
+
+Pros/cons of doing this:
+
+*   Simpler for users: users would be able to install a complete `jax` on CPU
+    with `pip install jax`. Provided we are able to get a size exception for our
+    GPU wheels, GPU wheels could be uploaded to Pypi and a GPU jax installation
+    could be installed with `pip install jax jaxlib-cuda-11-1`, without the need
+    for
+    [local versioning schemes](https://github.com/google/jax#pip-installation)
+    like `jaxlib+cuda110`that are not compatible with Pypi distribution.
+*   Reduced build times: `jaxlib` takes a long time to build; for example it
+    requires building XLA and LLVM. At present we cannot share build work
+    between, say, CUDA or Python versions, so to make a `jaxlib` release we
+    build ~12 different copies of XLA on Linux alone. By refactoring the
+    `jaxlib` wheels, we would remove the need to build the CPU support
+    repeatedly for each CUDA version, or the need to rebuild the GPU compiler
+    for each Python version.
+*   Splitting the wheels will reduce the total size of the binary distribution.
+    The largest `jaxlib` wheels at present are the CUDA wheels (~150MB each);
+    after refactoring we would no longer need to build one CUDA wheel per Python
+    version. This would reduce the size of our GPU wheel distribution by a
+    factor of 4, and it would make it easier to ship GPU wheels on Pypi.
```

Comment:
This thing of "build just one of X and not for every single python version" can also be accomplished by building one python-agnostic shared core library and then having the python bindings all in the same wheel that are minimal wrappers around that core.  See also https://github.com/tensorflow/tensorflow/pull/55941

Reply:
Indeed, this is the same idea just split across multiple wheels instead of having one wheel with multiple builds for different Python versions.

I think I prefer having a wheel split in here! It better describes to the Python packaging system what is happening and will degrade more gracefully when someone releases, say, a new Python version that we don't support. If we use separate packages, then `pip` will understand that it does not have a suitable wheel.

Reply:
> If we use separate packages, then pip will understand that it does not have a suitable wheel.

That's true.

Glad we're on the same basic page here, thanks.